src/ceph-volume/ceph_volume/devices/lvm/batch.py

```diff
@@ -593,46 +593,48 @@ def report_json(self) -> Dict[str, Any]:
             return {k: str(v) for k, v in self._get_osd_plan().items()}
 
 def get_physical_osds(devices: List[device.Device], args: argparse.Namespace) -> List[Batch.OSD]:
-    '''
-    Goes through passed physical devices and assigns OSDs
-    '''
-    data_slots = args.osds_per_device
-    if args.data_slots:
-        data_slots = max(args.data_slots, args.osds_per_device)
-    #rel_data_size = args.data_allocate_fraction / data_slots
-    #mlogger.debug('relative data size: {}'.format(rel_data_size))
+    """
+    Goes through passed physical devices and assigns OSDs.
+    """
+    data_slots = max(args.data_slots, args.osds_per_device) if args.data_slots else args.osds_per_device
     ret = []
+
     for dev in devices:
+        if not dev.available_lvm:
+            continue
+
+        total_dev_size = dev.vg_size[0]
+        dev_size = total_dev_size
         rel_data_size = args.data_allocate_fraction / data_slots
-        if dev.available_lvm:
-            total_dev_size = dev.vg_size[0]
-            dev_size = dev.vg_size[0]
 
-            if args.block_db_size is not None and not args.db_devices:
-                all_db_space = args.block_db_size * data_slots
-                dev_size -= all_db_space.b.as_int()
+        if args.block_db_size is not None and not args.db_devices:
+            all_db_space = args.block_db_size * data_slots
+            dev_size -= all_db_space.b.as_int()
 
-            abs_size = disk.Size(b=int(dev_size * rel_data_size))
-            mlogger.error(f'{dev_size} {abs_size} {rel_data_size}')
+        abs_size = disk.Size(b=int(dev_size * rel_data_size))
 
-            if not args.db_devices and args.block_db_size is not None:
-                rel_data_size = abs_size / disk.Size(b=total_dev_size)
+        if not args.db_devices and args.block_db_size is not None:
+            rel_data_size = abs_size / disk.Size(b=total_dev_size)
+
+        free_size = dev.vg_free[0]
+
+        for _ in range(args.osds_per_device):
+            if abs_size.b > free_size:
+                break
+
+            free_size -= abs_size.b
```

Comment:
is `b` in `abs_size.b` meant to be bytes?

Reply:
yes